spot_ros2_control/hardware/spot_hardware_interface.cpp

```diff
@@ -0,0 +1,373 @@
+      RCLCPP_FATAL(rclcpp::get_logger("SpotHardware"), "Joint '%s' have %s command interfaces found. '%s' expected.",
+                   joint.name.c_str(), joint.state_interfaces[0].name.c_str(), hardware_interface::HW_IF_EFFORT);
+      return hardware_interface::CallbackReturn::ERROR;
+    }
+  }
+  return hardware_interface::CallbackReturn::SUCCESS;
+}
+
+hardware_interface::CallbackReturn SpotHardware::on_configure(const rclcpp_lifecycle::State& /*previous_state*/) {
+  // reset values always when configuring hardware
+  for (uint i = 0; i < hw_states_.size(); i++) {
+    hw_states_.at(i) = 0;
+    hw_commands_.at(i) = 0;
+  }
+  return hardware_interface::CallbackReturn::SUCCESS;
+}
+
+std::vector<hardware_interface::StateInterface> SpotHardware::export_state_interfaces() {
+  std::vector<hardware_interface::StateInterface> state_interfaces;
+  for (uint i = 0; i < info_.joints.size(); i++) {
+    const auto joint = info_.joints.at(i);
+    state_interfaces.emplace_back(hardware_interface::StateInterface(joint.name, hardware_interface::HW_IF_POSITION,
+                                                                     &hw_states_[interfaces_per_joint_ * i]));
+    state_interfaces.emplace_back(hardware_interface::StateInterface(joint.name, hardware_interface::HW_IF_VELOCITY,
+                                                                     &hw_states_[interfaces_per_joint_ * i + 1]));
+    state_interfaces.emplace_back(hardware_interface::StateInterface(joint.name, hardware_interface::HW_IF_EFFORT,
+                                                                     &hw_states_[interfaces_per_joint_ * i + 2]));
+  }
+  return state_interfaces;
+}
+
+std::vector<hardware_interface::CommandInterface> SpotHardware::export_command_interfaces() {
+  std::vector<hardware_interface::CommandInterface> command_interfaces;
+  for (uint i = 0; i < info_.joints.size(); i++) {
+    const auto joint = info_.joints.at(i);
+    command_interfaces.emplace_back(hardware_interface::CommandInterface(joint.name, hardware_interface::HW_IF_POSITION,
+                                                                         &hw_commands_[interfaces_per_joint_ * i]));
+    command_interfaces.emplace_back(hardware_interface::CommandInterface(joint.name, hardware_interface::HW_IF_VELOCITY,
+                                                                         &hw_commands_[interfaces_per_joint_ * i + 1]));
+    command_interfaces.emplace_back(hardware_interface::CommandInterface(joint.name, hardware_interface::HW_IF_EFFORT,
+                                                                         &hw_commands_[interfaces_per_joint_ * i + 2]));
+  }
+  return command_interfaces;
+}
+
+hardware_interface::CallbackReturn SpotHardware::on_activate(const rclcpp_lifecycle::State& /*previous_state*/) {
+  // command and state should be equal when starting
+  for (uint i = 0; i < hw_states_.size(); i++) {
+    hw_commands_[i] = hw_states_[i];
+  }
```

Comment:
```suggestion
  hw_commands_.assign(hw_states_.begin(), hw_states_.end());
```

Should technically be faster and is very likely a memcpy internally